Content-Type gets wrongfully added to GET request after 303 redirects

sysrqb opened the pull request:
#### 9b8c53bed4f5ca3e81bc07fa1aed2b0d9dd6ae1a
<pre>
Content-Type gets wrongfully added to GET request after 303 redirects
<a href="https://bugs.webkit.org/show_bug.cgi?id=258720">https://bugs.webkit.org/show_bug.cgi?id=258720</a>
rdar://111899420

Reviewed by Chris Dumez.

In 252713@main we began copying the Content-Type header from the original
request if we receive a redirect with a 303 response status. However, we may be
handling a subsequent redirect, and in some cases the Content-Type header is
removed from redirect requests. For example, if we transition from a POST to a
GET request, Fetch 4.4.12 [0] says that the Content-Type header should be
deleted. WebKit currently follows this specification, but we re-introduce the
header on a second redirect because we consult the first request instead of the
previous request.

This patch introduces NetworkDataTask::m_previousRequest to keep track of the
previous request. This is duplicate information that is stored in NetworkLoad,
as well, but I&apos;m leaving merging these two copies for a follow-up patch.

[0] <a href="https://fetch.spec.whatwg.org/#http-redirect-fetch">https://fetch.spec.whatwg.org/#http-redirect-fetch</a>

* LayoutTests/imported/w3c/web-platform-tests/xhr/resources/redirect.py:
(main):
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-redirect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-redirect.htm:
* Source/WebCore/platform/network/ResourceHandleInternal.h:
* Source/WebCore/platform/network/mac/ResourceHandleMac.mm:
(WebCore::ResourceHandle::willSendRequest):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkLoad.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST):

Canonical link: <a href="https://commits.webkit.org/266761@main">https://commits.webkit.org/266761@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3605a618731840480ff3689a2399641e793836aa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14651 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14961 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15309 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16399 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13835 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14788 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17478 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/15044 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16511 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14832 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15340 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12451 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17133 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12623 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13219 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20229 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13702 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13385 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16630 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13934 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/11793 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13228 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3540 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17566 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13778 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez reviewed (changes requested):
This is breaking compatibility with other browsers in at least some aspects.

Consider this test change:
```
diff --git a/LayoutTests/imported/w3c/web-platform-tests/xhr/send-redirect.htm b/LayoutTests/imported/w3c/web-platform-tests/xhr/send-redirect.htm
index 16b3231e25b4..6e4ca3d4f89d 100644
--- a/LayoutTests/imported/w3c/web-platform-tests/xhr/send-redirect.htm
+++ b/LayoutTests/imported/w3c/web-platform-tests/xhr/send-redirect.htm
@@ -9,28 +9,45 @@
   <body>
     <div id="log"></div>
     <script>
-      function redirect(code) {
-        var test = async_test(document.title + " (" + code + ")")
+      function redirect(code, method) {
+        var test = async_test(document.title + " (" + code + ", " + method + ")")
         test.step(function() {
           var client = new XMLHttpRequest()
           client.onreadystatechange = function() {
             test.step(function() {
               if(client.readyState == 4) {
-                assert_equals(client.getResponseHeader("x-request-method"), "GET")
-                assert_equals(client.getResponseHeader("x-request-content-type"), "application/x-pony")
+                let expected_method = method;
+                if (method == "POST" && (code == "301" || code == "302" || code == "303"))
+                  expected_method = "GET";
+                assert_equals(client.getResponseHeader("x-request-method"), expected_method)
+                let expected_content_type = "application/x-pony";
+                if (method == "POST" && (code == "301" || code == "302" || code == "303"))
+                  expected_content_type = "NO";
+                assert_equals(client.getResponseHeader("x-request-content-type"), expected_content_type);
                 test.done()
               }
             })
           }
-          client.open("GET", "resources/redirect.py?location=content.py&code=" + code)
+          client.open(method, "resources/redirect.py?location=content.py&code=" + code)
           client.setRequestHeader("Content-Type", "application/x-pony")
           client.send(null)
         })
       }
-      redirect("301")
-      redirect("302")
-      redirect("303")
-      redirect("307")
+      redirect("301", "GET");
+      redirect("302", "GET");
+      redirect("303", "GET");
+      redirect("307", "GET");
+      redirect("308", "GET");
+      redirect("301", "POST");
+      redirect("302", "POST");
+      redirect("303", "POST");
+      redirect("307", "POST");
+      redirect("308", "POST");
+      redirect("301", "HEAD");
+      redirect("302", "HEAD");
+      redirect("303", "HEAD");
+      redirect("307", "HEAD");
+      redirect("308", "HEAD");
     </script>
   </body>
 </html>
```

This updated test passes in Chrome, Firefox and shipping Safari. However, your patch regresses the `XMLHttpRequest: send() - Redirects (basics) (303, GET)` subtest.

cdumez commented on LayoutTests/imported/w3c/web-platform-tests/xhr/send-redirect.htm:
```diff
@@ -9,20 +9,28 @@
     <script>
-      function redirect(code) {
-        var test = async_test(document.title + " (" + code + ")")
+      function redirect(code, method="GET", redirectLocation="content.py") {
```
That's great, please make sure to upstream to WPT or it will get lost during the next re-sync.

sysrqb replied:
https://github.com/web-platform-tests/wpt/pull/41417

cdumez commented on Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
```diff
@@ -1798,3 +1798,60 @@ HTTPServer server({
         Util::spinRunLoop();
 }
+
+TEST(WKNavigation, Multiple303Redirects)
```
Note that in general, we prefer layout tests to API tests whenever possible. This is a case where a layout test was possible I believe.

cdumez reviewed (approved):
LGTM

sysrqb committed:
Content-Type gets wrongfully added to GET request after 303 redirects
https://bugs.webkit.org/show_bug.cgi?id=258720
rdar://111899420

Reviewed by Chris Dumez.

In 252713@main we began copying the Content-Type header from the original
request if we receive a redirect with a 303 response status. However, we may be
handling a subsequent redirect, and in some cases the Content-Type header is
removed from redirect requests. For example, if we transition from a POST to a
GET request, Fetch 4.4.12 [0] says that the Content-Type header should be
deleted. WebKit currently follows this specification, but we re-introduce the
header on a second redirect because we consult the first request instead of the
previous request.

This patch introduces NetworkDataTask::m_previousRequest to keep track of the
previous request. This is duplicate information that is stored in NetworkLoad,
as well, but I'm leaving merging these two copies for a follow-up patch.

[0] https://fetch.spec.whatwg.org/#http-redirect-fetch

* LayoutTests/imported/w3c/web-platform-tests/xhr/resources/redirect.py:
(main):
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-redirect-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/xhr/send-redirect.htm:
* Source/WebCore/platform/network/ResourceHandleInternal.h:
* Source/WebCore/platform/network/mac/ResourceHandleMac.mm:
(WebCore::ResourceHandle::willSendRequest):
* Source/WebKit/NetworkProcess/NetworkDataTask.h:
* Source/WebKit/NetworkProcess/NetworkLoad.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::willPerformHTTPRedirection):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/Navigation.mm:
(TEST):

Canonical link: https://commits.webkit.org/266761@main
